### ISSUE TITLE
PubSub: Overwrite an item which has been published by an old (unregistered) user

### ISSUE THREAD

**Issue by ejbravo** [processone/ejabberd#2129]
> What version of ejabberd are you using?

17.11

> What operating system (version) are you using?

CentOS 7

> How did you install ejabberd (source, package, distribution)?

RPM package

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behaviour? What was the expected result?

In the latest ejabberd version it is not allowed to overwrite an item from a PubSub node once the original publisher is removed (unregistered). The result of this operation is a "forbidden" error and the only way to overwrite that "deprecated" item is removing it as the owner previously.

Overwriting data of removed users was possible in previous versions like 17.01 and it is the expected behaviour. Why did this change? Thanks in advance.

This issue has relation to #2067 ticket which has already been solved (thanks)

**Comment by weiss** [comment]
> Overwriting data of removed users was possible in previous versions like 17.01 and it is the expected behaviour. Why did this change?

Anyone permitted to publish on a node was also able to override items published by others.  This behaviour was not expected and therefore changed.

I agree that node *owners* should always be permitted to override/remove items, though (regardless of whether the publisher's account was removed - which you can only check for *local* accounts anyway).

**Comment by Emilio-Gonzalez** [comment]
I'm having the same problem.

Why anyone permitted to publish on a node is not able to override items published by others in the new v17.11? Why this behaviour was not expected?
In my opinion it is an interesting functionality and I use it often.

The XEP-0060: Publish-Subscribe says at 7.1.3 Error Cases (Publisher):
_Note: If a publisher publishes an item with an Item ID and the ItemID matches that of an existing item, the pubsub service MUST NOT fail the publication but instead MUST overwrite the existing item and generate a new event notification (i.e., re-publication is equivalent to modification)._

And now instead of overwrite the item the service returns a <forbidden/> error, **Insufficient Privileges**

Is not possible to get back to the previous behaviour ?
I think it is much more useful.

**Comment by weiss** [comment]
> Why anyone permitted to publish on a node is not able to override items published by others in the new v17.11?

Assume GitHub was using a PubSub node for our discussion.  Would you expect anyone to be able to override or delete your comments just because anyone is permitted to add a comment?

> XEP-0060

Yes, the XEP is not very explicit on this case.  I would assume the author of your quote didn't have the case in mind where the publisher of the overriding item is not the publisher of the original item.  This is more obvious if you look at the wording for the [success case][1]:

> Note: If *the publisher* previously published an item with the same ItemID, successfully processing the request means that the service MUST overwrite the old item with the new item ... [emphasis added]

[1]: https://xmpp.org/extensions/xep-0060.html#publisher-publish-success

**Comment by Emilio-Gonzalez** [comment]
> Assume GitHub was using a PubSub node for our discussion. Would you expect anyone to be able to override or delete your comments just because anyone is permitted to add a comment?

Good point.

Probably not everyone should be able to override or delete my comments, but what about administrators, webmasters, high-level users, etc.
Now, that's not possible? No one can edit my comment?
Needs the master use my user to change the comments? Makes sense?

**Comment by ejbravo** [comment]
I agree with @Emilio-Gonzalez and its interpretation of [XEP-0060](https://xmpp.org/extensions/xep-0060.html#publisher-publish). But anyway, assuming that it is not possible to edit an item which has been published by an author who is not the original publisher (this new interpretation for 17.11 version), what can I do with those items that were published by a removed user? is the only option removing them using the admin account to update them? (remove and publish again) 

Let me explain with more details my case:

I have created a node with the next configuration:
- _access_model_: open (Anyone may subscribe and retrieve items)
- _publish_model_: open (Anyone may publish)
- _persist_items_: true (Whether to persist items to storage)
- _purge_offline_: true (Whether to purge all items when the relevant publisher goes offline)

This issue is caused because of a weird interpretation of this last point, it seems that when a user is removed, Pubsub does not understand that this user is offline and all of his items keep alive on the node. Is it not obvious that an unregistered user is offline? When I try to edit these items the new users receive the "forbidden" error. This has no sense for me. Is could be possible to add another boolean option in Pubsub node configuration to permit overwriting of published items?

**Comment by Emilio-Gonzalez** [comment]
Has anything changed in the v18.01?

I need to do more tests, but it seems possible to overwrite an item value using a different user from the owner.
Not possible to retract the item.

Is it right?

**Comment by cromain** [comment]
this fix improves the situation, by purging items of removed user.
now as @weiss said, current implementation should be kept. as long a user is registered he should be the only one (with node owner) able to alter its own items.
so i guess we can close this issue. @weiss ?

**Comment by Emilio-Gonzalez** [comment]
As I remember, the user was the only one able to alter its own items. But the node owner wasn't. And that's one of the issue keys.

The other one was that user items weren't purged when the user was removed.
If the fix improves the item purging of removed user, it's ok.

But it's needed the node owner will be able to alter all items in that node whenever.

**Comment by weiss** [comment]
> this fix improves the situation, by purging items of removed user.

Sounds like a good solution for most use cases (I want this behavior on my own servers).  Not sure *everyone* wants this, but we'll see :smile:

**Comment by ejbravo** [comment]
I'm not sure if I have understood this last point.

> this fix improves the situation, by purging items of the removed user.

Will be the behaviour "purging items of removed users" the solution in every case? I mean, when the node configuration   says in its options: 

- purge_offline: true (Whether to purge all items when the relevant publisher goes offline)

That solution is the perfect solution. However, when the option purge_offline is not "true", will be the behaviour the same and will be the items of the removed user purged as well?

**Comment by Emilio-Gonzalez** [comment]
Sorry, I think I was wrong. The node owner is able to alter the items.

**Comment by cromain** [comment]
@ejbravo when purge_offline is false, items are not removed.
the last fix is about removing items of user being unregistered.

**Comment by ServiceTo** [comment]
@Emilio-Gonzalez in my setup, the issue I'm having is that everyone in the node is an owner but only the original publisher of the item can update it.  I need them all to be able to update anything. (presence info for extensions can come from multiple servers)

I'm getting 
```
<error code='403' type='auth'>
  <forbidden xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/>
</error>
``` 
on servers who didn't initially create the state for that extension but should be able to update it.

**Comment by Emilio-Gonzalez** [comment]
Well, then I was not so wrong.
@ServiceTo Could you give us more details about your issue and your configuration?

**Comment by ServiceTo** [comment]
I'm using XMPP PubSub for Asterisk presence management for distributed call queues.  I previously used Tigase but ejabberd looked pretty cool and not Java so I figured I'd go there.  The only thing I'm missing in my setup is that each user needs to be able to write the latest state of an extension and currently only the first user to create that extension's state in the pubsub node is allowed to update it.

I've made all of the users owners of the node but still the only user that can update the node item is the one who created it initially, which happens to be okay right this second but should I be required swap out an instance, will be broken.
